std/range/package.d

```diff
@@ -1228,6 +1231,382 @@ unittest
+    result is a range that offers the $(D front), $(D popFront), and $(D
+    empty) primitives. If all input ranges offer random access and $(D
+    length), $(D choose) offers them as well.
+
+    Params:
+        index = which range to choose, must be less than the number of ranges
+        rs = tuple of 1 or more ranges
+
+    Returns:
+        The indexed range. If rs consists of only one range,
+        the return type is an alias of that range's type.
+ */
+auto choose(Ranges...)(size_t index, Ranges rs)
+if (Ranges.length > 0 &&
+    allSatisfy!(isInputRange, staticMap!(Unqual, Ranges)) &&
+    !is(CommonType!(staticMap!(ElementType, staticMap!(Unqual, Ranges))) == void))
+{
+    static if (Ranges.length == 1)
+    {
+        assert(index == 0);
+        return rs[0];
+    }
+    else
+    {
+        static struct Result
+        {
+        private:
+            alias R = staticMap!(Unqual, Ranges);
+            alias RvalueElementType = CommonType!(staticMap!(.ElementType, R));
+            private template sameET(A)
+            {
+                enum sameET = is(.ElementType!A == RvalueElementType);
+            }
+
+            enum bool allSameType = allSatisfy!(sameET, R);
+
+            static if (allSameType)
+            {
+                alias ref RvalueElementType ElementType;
+            }
+            else
+            {
+                alias ElementType = RvalueElementType;
+            }
+            static if (allSameType && allSatisfy!(hasLvalueElements, R))
+            {
+                static ref RvalueElementType fixRef(ref RvalueElementType val)
+                {
+                    return val;
+                }
+            }
+            else
+            {
+                static RvalueElementType fixRef(RvalueElementType val)
+                {
+                    return val;
+                }
+            }
+
+            // Create a tagged union for the Ranges
+            size_t index;               // the tag
+            union U { R fields; }
+            U source;
+
+        public:
+            this(size_t index, R input)
+            {
+                this.index = index;
+                foreach (i, v; input)
+                {
+                    if (i == index)
+                    {
+                        source.fields[i] = v;
+                        return;
+                    }
+                }
+                assert(0);
+            }
+
+            import std.typetuple : anySatisfy;
+
+            static if (allSatisfy!(isInfinite, R))
+            {
+                // Propagate infiniteness.
+                enum bool empty = false;
+            }
+            else
+            {
+                @property bool empty()
+                {
+                    foreach (i, Unused; R)
+                    {
+                        if (index == i)
+                            return source.fields[i].empty;
+                    }
+                    assert(0);
+                }
```

Comment:
This pattern loop/if/assert(0) is common through the code. Is it susceptible to be optimized well for the frequent case Ranges.length == 2? I see two comparisons here, more than needed. The bool case would go:

```D
@property bool empty()
{
  return index ? source.fields[1].empty : source.fields[0].empty;
}
```